### ISSUE TITLE
Refactoring

### ISSUE THREAD

**Issue by Superhepper** [parallaxsecond/rust-tss-esapi#87]
The files in the library is getting way to big. It is time to try to do something about that. And while we are at it we should probably come up with some better name and structure. This is just something I go and think about while doing other stuff so I have no proposal yet how to do it. But I am putting here so that we can start a discussion about it.

One thing I can say that personally like is that lib and mod files does not contain any code. They just contain documentation, settings, clippy-settings, use statements and mod statements.

**Comment by hug-dev** [comment]
Definitely agree with refactoring and your two points.

**Comment by ionut-arm** [comment]
Yeah, unfortunately when I wrote a lot of the code, especially the utils stuff, I didn't pay much attention to how it would scale 😞 

The suggestion sounds pretty good, we can have a big `context` module that splits up the `Context` into the areas you mentioned. For the utils maybe something similar, or maybe we can draw some inspiration from the way the spec (especially the one on structures) is split.

**Comment by Superhepper** [comment]
I will begin a 'phase 1' of this work.
This will involve the following:
1. Move context out of the lib file.
2. Move pcr selection stuff out of mod in utils.

**Comment by Superhepper** [comment]
For phase 2 I am thinking if doing the following:
- Move tickets to structures.
- Create new constants folder.
- Rename constants.rs to tss.rs and move it into constants.
- Add tags to constants folder.

**Comment by Superhepper** [comment]
Only one thing left to do now are the things in utils. And that is to move the remaining items out of utils/mod.rs and into their correct locations. I guess there are things in there that are not tied to any TPM structures and those I think should remain there
though in their own files.

Any one who feels they got the hang of those structures feel free to do this. I had some ideas regarding some of the things. I have been thinking that for consistency reasons there should be representations for TPMT_PUBLIC and TPM2B_PUBLIC(e.t.c.).

And if we look at the specification it should be something like:
```
|-- src
    |-- complex (key/object Complex)
        |-- mod.rs
        |-- public
            |--mod.rs
            |--tagged.rs (TPMT_PUBLIC, struct TaggedPublic)
            |--public.rs (Not sure about the name could be called buffer.rs as well. TPM2B_PUBLIC, struct Public).
```

Take these suggestions with a barrel of salt. As I said I have only been pondering about it. But not really committed any energy to it hehe.

**Comment by ionut-arm** [comment]
I think we've gotten to a fairly nice structure, but there are still some inconsistencies. For version 6.0.0 we should aim for our modules to follow the chapters/sections in part 2 (Structures) of the spec.

`src/structures`, for example, should (only) contain the types and structures in chapter 10 - `Structure definitions`, with a module for each sub-section and so on. At the moment there are a few types in `structures` that don't belong there.

Whenever this is picked up again, a full mapping should be agreed-upon for the top-level modules at least, with their name and a brief description.

**Comment by Superhepper** [comment]
I am sure we had a discussion regarding this at some point(can't remember if it was on slack or in a PR). Were we agreed that there were two ways to structure the code.

1. Structure it in accordance with the specification.
2. Or structure it with regard to the prefixes of the types.

Perhaps it is possible to combine them?

**Comment by ionut-arm** [comment]
I think it is possible to combine them, though I'm not sure duplicating the types would make it easier to follow - there are already a lot of types with really similar names.

I don't know how a user of the crate would go about finding out how to do something: would the start from the spec or from the crate documentation? Because if you start from the crate docs, for each method on `Context` you get links to the types you need to use, which also includes docs for those types too (I guess this is another reason to have native types everywhere)

**Comment by Superhepper** [comment]
I think how some one tries to find the types depends on where they come from. Like a beginner would probably click around in the documentation and for that person the usage of native types that are well documented are really important.

Some one more experienced, that might have been writing TPM related code in C, and is well familiar with the specifications might look at modules sections and expects to find names and structuring that corresponds very well to the specification.

**Comment by ionut-arm** [comment]
> Some one more experienced, that might have been writing TPM related code in C, and is well familiar with the specifications might look at modules sections and expects to find names and structuring that corresponds very well to the specification.

True, but even then, given that we've changed at least part of the spec to look more Rusty, they _should_ be starting with some operations to perform, right? I don't think any of the types would be useful to look at on their own
